### PR TITLE
feat: use volumes for snapshot generation

### PR DESCRIPTION
**Pull request by lemmih**
**Summary of changes**
Changes introduced in this pull request:
- Use volumes when exporting snapshots to avoid running out of disk space.
- Use cron+flock to schedule the snapshot scripts (more robust and restarts automatically on reboots)
- Switch from s3fs to s3cmd (s3fs isn't 100% supported and has weird issues)
- Generate new snapshots every hour instead of only once per day. This means several snapshots might be pruned every day.
- Update the snapshot pruning to not delete the snapshot that used to be the most-recent snapshot. This snapshot might still be accessed and s3 downloads are not atomic. The snapshot will still be pruned later on, though. By then, any download should have been completed.
- Drop the calibnet/mainnet distinction. Both snapshots will be computed on the same server.

**Reference issue to close (if applicable)**

Closes 

**Other information and links**

### CODE REVIEW

File: daily_snapshot_terraform/modules/daily_snapshot/service/daily_snapshot.rb [hunk 1]
```diff
@@ -13,47 +14,43 @@
...
 
-loop do
-  client = SlackClient.new CHANNEL, SLACK_TOKEN
+client = SlackClient.new CHANNEL, SLACK_TOKEN
 
-  # Find the snapshot with the most recent modification date
-  latest = Dir.glob(File.join(SNAPSHOTS_DIR, '/*.car')).max_by { |f| File.mtime(f) }
+# Find the snapshot with the most recent modification date
+all_snapshots = list_snapshots(CHAIN_NAME, BUCKET, ENDPOINT)
+if !all_snapshots.empty?
```
Comment (LesnyRumcajs) [inline daily_snapshot_terraform/modules/daily_snapshot/service/daily_snapshot.rb:33]:
```suggestion
unless all_snapshots.empty?
```

File: daily_snapshot_terraform/modules/daily_snapshot/service/daily_snapshot.rb [hunk 2]
```diff
@@ -13,47 +14,43 @@
...
+  latest = all_snapshots[0]
```
Comment (LesnyRumcajs) [inline daily_snapshot_terraform/modules/daily_snapshot/service/daily_snapshot.rb:34]:
```suggestion
  latest = all_snapshots.first
```

File: daily_snapshot_terraform/modules/daily_snapshot/service/daily_snapshot.rb [hunk 3]
```diff
@@ -13,47 +14,43 @@
...
 
-    # Sync and export snapshot
-    snapshot_uploaded = system("bash upload_snapshot.sh #{CHAIN_NAME} #{latest} > #{LOG_EXPORT} 2>&1")
+  # Update our list of snapshots
+  all_snapshots = list_snapshots(CHAIN_NAME, BUCKET, ENDPOINT)
 
-    if snapshot_uploaded
+  if snapshot_uploaded
+    # If this is the first new snapshot of the day, send a victory message to slack
+    latest = all_snapshots[0]
```
Comment (LesnyRumcajs) [inline daily_snapshot_terraform/modules/daily_snapshot/service/daily_snapshot.rb:44]:
```suggestion
    latest = all_snapshots.first
```

File: daily_snapshot_terraform/modules/daily_snapshot/service/init.sh [hunk 4]
```diff
@@ -1,27 +1,27 @@
...
+    --multipart-chunk-size-mb=4096 > ~/.s3cfg
+
+## Setup volume
+mkdir forest_db
+mount -o defaults,nofail,discard,noatime /dev/disk/by-id/scsi-0DO_Volume_snapshot-gen-storage forest_db
+chmod 777 forest_db
+
+# Setup cron jobs
+cp calibnet_cron_job /etc/cron.hourly/
+cp mainnet_cron_job /etc/cron.hourly/
```
Comment (LesnyRumcajs) [inline daily_snapshot_terraform/modules/daily_snapshot/service/init.sh:27]:
```suggestion
cp mainnet_cron_job calibnet_cron_job /etc/cron.hourly/
```

File: daily_snapshot_terraform/modules/daily_snapshot/service/list_snapshots.rb [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+  snapshot_list = []
+
+  output.each_line do |line|
+    line.match(ls_format) { |m|
+      file = m.captures[0]
+      file.match(snapshot_format) { |m| 
+        url = "https://#{bucket}.#{endpoint}/#{chain_name}/#{file}"
+        url_s3 = "s3://#{bucket}/#{chain_name}/#{file}"
+        snapshot = Snapshot.new m[:network], m[:date].to_date, Integer(m[:height]), m[:compressed], file, url_s3, url
```
Comment (LesnyRumcajs) [inline daily_snapshot_terraform/modules/daily_snapshot/service/list_snapshots.rb:27]: I think `to_i` would work as well, but not sure (for the `Integer`)

File: daily_snapshot_terraform/modules/daily_snapshot/service/mainnet_cron_job [hunk 6]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
+cd "$BASE_FOLDER" || exit
```
Comment (LesnyRumcajs) [inline daily_snapshot_terraform/modules/daily_snapshot/service/mainnet_cron_job:2]: it's not an error?

File: daily_snapshot_terraform/modules/daily_snapshot/service/snapshots_prune.rb [hunk 7]
```diff
@@ -22,47 +24,45 @@ def add?(entry)
...
-  File.delete checksum if checksum.file?
-  File.delete snapshot
+  snapshots
+    # ignore the first snapshot to keep a buffer of two recent snapshots.
+    # this makes it less likely that we delete a snapshot that is being downloaded.
+    .drop(1)
+    # keep snapshots (ie reject) if they fit in a bucket while also having a unique date
+    .reject  { |f| day_unique_bucket.add? f and buckets.any? { |bucket| bucket.add? f } }
+    # delete all snapshots that weren't rejected or dropped
+    .each    { |f| f.delete }
```
Comment (LesnyRumcajs) [inline daily_snapshot_terraform/modules/daily_snapshot/service/snapshots_prune.rb:67]:
```suggestion
    .each(&:delete)
```

File: daily_snapshot_terraform/modules/daily_snapshot/service/mainnet_cron_job (hunk 6 above)
Comment (lemmih) [reply daily_snapshot_terraform/modules/daily_snapshot/service/mainnet_cron_job:2]: Without an error code, `exit` uses the error status of the last executed command.

File: daily_snapshot_terraform/modules/daily_snapshot/service/snapshots_prune.rb (hunk 7 above)
Comment (lemmih) [reply daily_snapshot_terraform/modules/daily_snapshot/service/snapshots_prune.rb:67]: Neat.

File: daily_snapshot_terraform/modules/daily_snapshot/service/calibnet_cron_job [hunk 8]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
+cd "$BASE_FOLDER" || exit
+# shellcheck source=/dev/null
+source .bashrc
```
[the remark this answers is not in the document]
Comment (lemmih) [reply daily_snapshot_terraform/modules/daily_snapshot/service/calibnet_cron_job:4]: I'll give it a try.

[the remark this answers is not in the document]
Comment (lemmih) [reply daily_snapshot_terraform/modules/daily_snapshot/service/calibnet_cron_job:4]: Nope, doesn't work.

File: daily_snapshot_terraform/modules/daily_snapshot/main.tf [hunk 9]
```diff
@@ -71,20 +81,27 @@ resource "digitalocean_droplet" "forest" {
...
       # Set required environment variables
-      "export AWS_ACCESS_KEY_ID=${var.AWS_ACCESS_KEY_ID}",
-      "export AWS_SECRET_ACCESS_KEY=${var.AWS_SECRET_ACCESS_KEY}",
-      "export SLACK_API_TOKEN=${var.slack_token}",
-      "export SLACK_NOTIF_CHANNEL=${var.slack_channel}",
-      "export SNAPSHOT_BUCKET=${var.snapshot_bucket}",
-      "export SNAPSHOT_ENDPOINT=${var.snapshot_endpoint}",
-      # Run init script in the background
-      "nohup sh ./init.sh ${var.chain} &",
+      "echo 'export AWS_ACCESS_KEY_ID=\"${var.AWS_ACCESS_KEY_ID}\"' >> .bashrc",
```
[the remark this answers is not in the document]
Comment (lemmih) [reply daily_snapshot_terraform/modules/daily_snapshot/main.tf]: Moved to `.forest_env`.

File: daily_snapshot_terraform/modules/daily_snapshot/service/mainnet_cron_job (hunk 6 above)
Comment (lemmih) [reply daily_snapshot_terraform/modules/daily_snapshot/service/mainnet_cron_job:2]: I didn't either! shellcheck taught me. :D

File: daily_snapshot_terraform/modules/daily_snapshot/service/calibnet_cron_job (hunk 8 above)
[the remark this answers is not in the document]
Comment (LesnyRumcajs) [reply daily_snapshot_terraform/modules/daily_snapshot/service/calibnet_cron_job:4]: sad

**Review by LesnyRumcajs — APPROVED** [review]

**Pull request merged**